src/compas_fab/backends/pybullet/client.py

```diff
@@ -175,44 +184,103 @@ def load_robot(self, urdf_file):
         """
         robot_model = RobotModel.from_urdf_file(urdf_file)
         robot = Robot(robot_model, client=self)
+        robot.attributes['pybullet'] = {}
+        if resource_loaders:
+            robot_model.load_geometry(*resource_loaders)
+            self.cache_robot(robot)
+        else:
+            robot.attributes['pybullet']['cached_robot'] = robot.model
+            robot.attributes['pybullet']['cached_robot_filepath'] = urdf_file
 
+        urdf_fp = robot.attributes['pybullet']['cached_robot_filepath']
+
+        self._load_robot_to_pybullet(urdf_fp, robot)
+
+        return robot
+
+    def _load_robot_to_pybullet(self, urdf_file, robot):
+        cached_robot = robot.attributes['pybullet']['cached_robot']
         with redirect_stdout(enabled=not self.verbose):
             pybullet_uid = pybullet.loadURDF(urdf_file, useFixedBase=True,
                                              physicsClientId=self.client_id,
                                              flags=pybullet.URDF_USE_SELF_COLLISION)
-            robot.attributes['pybullet_uid'] = pybullet_uid
+            cached_robot.attr['uid'] = pybullet_uid
 
-        self._add_ids_to_robot_joints(robot)
-        self._add_ids_to_robot_links(robot)
-        return robot
+        self._add_ids_to_robot_joints(cached_robot)
+        self._add_ids_to_robot_links(cached_robot)
 
-    def _add_ids_to_robot_joints(self, robot):
-        joint_ids = self._get_joint_ids(robot.attributes['pybullet_uid'])
+    @staticmethod
+    def ensure_cached_robot(robot):
+        if not robot.attributes['pybullet'].get('robot_cached_in_tmpdir'):
+            raise Exception(
+                'This method is only callable once the robot has been cached.')
+
+    def cache_robot(self, robot):
+        robot.ensure_geometry()
+        # write meshes to cache
+        address_dict = {}
+        # must work with given robot.model here because it has the geometry loaded
+        for link in robot.model.links:
+            for element in itertools.chain(link.visual, link.collision):
+                shape = element.geometry.shape
+                if isinstance(shape, MeshDescriptor):
+                    mesh = shape.geometry
+                    mesh_file_name = str(mesh.guid) + '.stl'
```

Comment:
Here I guess we have information loss if the meshes are DAE files, right? It's ok not to solve it in this PR, but maybe add a `TODO` note.

Reply:
do you mean that the pybullet representation of a mesh loaded from a DAE file might lose its textures or whatnot?  Erwin [says](https://github.com/bulletphysics/bullet3/issues/2175#issuecomment-476467198) loading dae files is not well supported, and it seems quite a lot of issues have been logged around pybullet and dae.  so i could change to obj rather than stl and lose a little less...

Reply:
Ah! Didn't know that. Nah, then it's fine!